tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/store/AbstractDBStore.java

```diff
@@ -2228,6 +2244,257 @@ private void setPartitionKeys(int pid, PartitionDescProto.Builder partitionDesc)
+        pstmt.setInt(currentIndex, tableId);
+        currentIndex++;
+
+        for (Pair<Type, Object> parameter : filter.getParameters()) {
+          switch (parameter.getFirst()) {
+            case BOOLEAN:
+              pstmt.setBoolean(currentIndex, (Boolean)parameter.getSecond());
+              break;
+            case INT8:
+              pstmt.setLong(currentIndex, (Long) parameter.getSecond());
+              break;
+            case FLOAT8:
+              pstmt.setDouble(currentIndex, (Double) parameter.getSecond());
+              break;
+            case DATE:
+              pstmt.setDate(currentIndex, (Date) parameter.getSecond());
+              break;
+            case TIMESTAMP:
+              pstmt.setTimestamp(currentIndex, (Timestamp) parameter.getSecond());
+              break;
+            case TIME:
+              pstmt.setTime(currentIndex, (Time) parameter.getSecond());
+              break;
+            default:
+              pstmt.setString(currentIndex, (String) parameter.getSecond());
+              break;
+          }
+          currentIndex++;
+        }
+      }
+
+      res = pstmt.executeQuery();
+
+      while (res.next()) {
+        PartitionDescProto.Builder builder = PartitionDescProto.newBuilder();
+
+        builder.setId(res.getInt(COL_PARTITIONS_PK));
+        builder.setPartitionName(res.getString("PARTITION_NAME"));
+        builder.setPath(res.getString("PATH"));
+        builder.setNumBytes(res.getLong(COL_PARTITION_BYTES));
+        builder.setNumFiles(res.getLong(COL_PARTITION_FILES));
+
+        partitions.add(builder.build());
+      }
+    } catch (TajoException se) {
+      throw new TajoInternalError(se);
+    } catch (SQLException se) {
+      throw new TajoInternalError(se);
+    } finally {
+      CatalogUtil.closeQuietly(pstmt, res);
+    }
+
+    return partitions;
+  }
+
+  /**
+   * Create a select statement and parameters for querying partitions and partition keys in CatalogStore.
+   *
+   * For example, consider you have a partitioned table for three columns (i.e., col1, col2, col3).
+   * Assume that an user gives a condition WHERE (col1 ='1' or col1 = '100') and col3 > 20.
+   * There is no filter condition corresponding to col2.
+   *
+   * Then, the sql would be generated as following:
+   *
+   *  SELECT A.PARTITION_ID, A.PARTITION_NAME, A.PATH FROM PARTITIONS A
+   *  WHERE A.TID = ?
+   *  AND A.PARTITION_ID IN (
+   *    SELECT T1.PARTITION_ID FROM PARTITION_KEYS T1
+   *    JOIN PARTITION_KEYS T2 ON T1.TID=T2.TID AND T1.PARTITION_ID = T2.PARTITION_ID AND T2.TID = ?
+   *    AND ( T2.COLUMN_NAME = 'col2' AND T2.PARTITION_VALUE IS NOT NULL )
+   *    JOIN PARTITION_KEYS T3 ON T1.TID=T3.TID AND T1.PARTITION_ID = T3.PARTITION_ID AND T3.TID = ?
+   *    AND ( T3.COLUMN_NAME = 'col3' AND T3.PARTITION_VALUE > ? )
+   *    WHERE T1.TID = ? AND ( T1.COLUMN_NAME = 'col1' AND T1.PARTITION_VALUE = ? )
+   *    OR ( T1.COLUMN_NAME = 'col1' AND T1.PARTITION_VALUE = ? )
```

Comment:
I also think that self-join on partition_keys is expensive way. But it might need more time to improve the select statement and according to the need, we would change partition schema. So, I want to handle this issue at another issue. What do you think about my opinion? For your information, you can see the performance of this method at https://github.com/apache/tajo/pull/624#issuecomment-127275276 .